### PR TITLE
add silu_and_mul_with_clamp kernel

### PR DESCRIPTION
**Pull request by yangulei**
## Summary

Adds the `silu_and_mul_with_clamp` XPU activation kernel (SwiGLU-OAI with input clamping, used by MiniMax-M3 MoE) and a bandwidth-optimized SYCL implementation.

Semantics (matches the CUDA reference and `torch.clamp`-based native path):

```
gate = input[:, :d];  up = input[:, d:]      # contiguous halves
gate_c = min(gate, limit)
up_c   = clamp(up, -limit, limit)
silu   = gate_c / (1 + exp(-alpha * gate_c))
out    = silu * (up_c + beta)                # alpha=1, beta=0 -> silu(gate)*up
```

## Implementation

- **`silu_and_mul_with_clamp_vec_kernel<T, ACT_FN, VEC>`** — the fast path used when `d % VEC == 0` (`VEC = 16/sizeof(T)`, i.e. 8×bf16/half, 4×float):
  - **16-byte aligned vector loads/stores** via `vllm::xpu::aligned_vec`, so each work-item moves a full cache-line-friendly chunk instead of scalar 2-byte elements.
  - **Flat grid-stride dispatch** over all `(token, chunk)` pairs (WG=256, grid capped at 2560 workgroups) so GPU occupancy is independent of the token count — small-batch shapes no longer launch just a handful of workgroups.
- The original scalar per-token kernel is retained as the fallback for `d % VEC != 0`, preserving correctness for arbitrary `d`.

## Performance (Intel Arc Pro B60 / BMG, bf16)

Profile-guided optimization on the MiniMax-M3 shape matrix. Effective bandwidth of the fused op (read 2·d + write d):

| Shape (num_tokens × d) | Before | After | Speedup |
|---|---:|---:|---:|
| **16384 × 768** (dominant, TritonExperts) | 262 GB/s | **390 GB/s** | **+49%** |
| 4096 × 768 | 293 | 757 | +158% |
| 2048 × 768 | 348 | 790 | +127% |

The dominant shape reaches **85.7% of the 456 GB/s theoretical peak** and **99.5% of the empirical streaming roofline** (a trivial same-traffic copy tops out at 393 GB/s), i.e. it is at the memory-bandwidth wall. ISA analysis confirmed the remaining ALU (clamp `cmp/sel`, sigmoid `exp`) is fully hidden behind memory latency, so no further ISA-level tuning helps.

Versus the PyTorch-native (`torch.clamp`/`sigmoid`/`mul`) path, the fused kernel is **~8.4× faster (geomean)** across the shape matrix and also more accurate (fp32 intermediates, single cast: err 0.125 vs 0.24 against an fp32 golden).

## Tests

`tests/test_silu_and_mul_with_clamp.py` validates the op against the native reference across `{half, bfloat16, float}`, multiple token counts, and both `d` divisible by the vector width (vectorized path) and `d = 17` (scalar fallback). Plus `opcheck` for schema/meta.

```
54 passed
```

### CODE REVIEW

**Review by jikunshang — APPROVED** [review]

**Comment by jikunshang** [comment]
please rebase and resolve conflicts thanks.

File: csrc/torch_bindings.cpp [hunk 1]
```diff
@@ -200,6 +200,14 @@ TORCH_LIBRARY_EXPAND(TORCH_EXTENSION_NAME, ops) {
       "-> ()");
   ops.impl("swiglustep_and_mul", torch::kXPU, &swiglustep_and_mul);
 
+  // SwiGLU (SiLU gate) with input clamping.
+  // out = silu(gate.clamp(max=limit)) * (up.clamp(+-limit) + beta), where
```
Comment (Liangliang-Ma) [inline csrc/torch_bindings.cpp]: let's keep torch bindings clean. please remove additional comments here.

File: tests/test_silu_and_mul_with_clamp.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+DTYPES = [torch.half, torch.bfloat16, torch.float]
+NUM_TOKENS = [7, 83, 2048]  # Arbitrary values for testing
+D = [512, 13824, 17]  # 17 exercises the scalar fallback (d % vec_size != 0)
+LIMITS = [7.0]
+PARAMS = [(1.0, 0.0), (1.702, 1.0)]  # (alpha, beta)
+SEEDS = [0]
+XPU_DEVICES = [
+    f"xpu:{i}" for i in range(1 if torch.xpu.device_count() == 1 else 2)
+]
```
[the remark this answers is not in the document]
Comment (yangulei) [reply tests/test_silu_and_mul_with_clamp.py]: Fixed in 2f1309a: `XPU_DEVICES` now uses `min(torch.xpu.device_count(), 2)`, so the list is empty on hosts without an XPU instead of generating `xpu:1`.

File: tests/test_silu_and_mul_with_clamp.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+@torch.inference_mode()
+def test_silu_and_mul_with_clamp(
+    num_tokens: int,
+    d: int,
+    limit: float,
+    params: tuple,
+    dtype: torch.dtype,
+    seed: int,
+    device: str,
+) -> None:
```
[the remark this answers is not in the document]
Comment (yangulei) [reply tests/test_silu_and_mul_with_clamp.py:56]: `reset_default_device` in `tests/conftest.py` is declared `@pytest.fixture(autouse=True)`, so it already wraps every test in this file and restores the default device — no leak. The device selection itself is still fixed in 2f1309a (see the next thread).

File: tests/test_silu_and_mul_with_clamp.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    seed_everything(seed)
+    torch.set_default_device(device)
+    alpha, beta = params
+    x = torch.randn(num_tokens, 2 * d, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (yangulei) [reply tests/test_silu_and_mul_with_clamp.py:62]: Fixed in 2f1309a: the test now follows `tests/test_layernorm.py` and does `torch.set_default_device("xpu")` + `torch.xpu.set_device(device)`, so the parametrized device is honoured. Verified with 120 passing cases across `xpu:0`/`xpu:1`.

File: csrc/activation.cpp [hunk 5]
```diff
@@ -714,3 +841,70 @@ void swiglustep_and_mul(
...
 }
+
+#define LAUNCH_SILU_AND_MUL_WITH_CLAMP(KERNEL, LIMIT, ALPHA, BETA)          \
+  int d = input.size(-1) / 2;                                               \
+  int64_t num_tokens = input.numel() / input.size(-1);                      \
+  at::DeviceGuard device_guard(input.device());                             \
+  auto& queue = vllm::xpu::vllmGetQueue();                                  \
+  if (num_tokens == 0) {                                                    \
+    return;                                                                 \
+  }                                                                         \
```
[the remark this answers is not in the document]
Comment (yangulei) [reply csrc/activation.cpp]: Fixed in 2f1309a: `LAUNCH_SILU_AND_MUL_WITH_CLAMP` now returns early on `input.numel() == 0` before `num_tokens` is computed, so a zero-sized last dimension can no longer divide by zero. Covered by the new `test_silu_and_mul_with_clamp_empty_input` case (shapes `(0, 512)` and `(4, 0)`).

File: csrc/torch_bindings.cpp (hunk 1 above)
Comment (yangulei) [reply csrc/torch_bindings.cpp]: Done — the extra comment block in `csrc/torch_bindings.cpp` is removed; only a single `// silu_and_mul_with_clamp` line remains, matching the neighbouring op registrations. The branch is also rebased onto latest `main` with conflicts resolved.
